wordpress-mobile/WordPress-iOS
# [Tooling] Migrate ShareExtension to use strings from app bundle

📝  This is part of project ref paaHJt-2Ib-p2 & paaHJt-2J8-p2
➕  This builds on top of https://github.com/wordpress-mobile/WordPress-iOS/pull/17630 and the `tooling/l10n-step2/ios13-widgets` branch.

## What's in this PR

Same stuff as what I did with the iOS14-style Widget and the 3 iOS13-style widgets in previous PRs:
1. Added `AppLocalizedStrings.swift` to the `WordPressShareExtension` target. Also had to add it to the `WordPressDraftActionExtension` target, because files which I migrated to `AppLocalizedString` in step 2 happen to also be part of that target too, so I needed it to make that target continue to compile too
2. Migrated all the usages of `NSLocalizedString` in any of the files that are part of that `WordPressShareExtension`'s "Compile Sources" build phase, to use `AppLocalizedString` instead, so that it looks up the strings in the app bundle(†)
3. Checked(‡) that all the keys and values in the `*.lproj/Localizable.strings` files that were in  the `WordPressShareExtension` target so fare were also present in the corresponding `WordPress/Resources/*.lproj/Localizable.strings`, with the same value, to ensure it was safe to delete them
4. Delete all the `WordPress/WordPressShareExtension/*.lproj/Localizable.strings` files from the disk and the `ShareExtension` target in Xcode

#### Note: Scripts I used to help me with steps 2 & 3

<details><summary>(†) Script to check that the keys in <tt>.strings</tt> files matched</summary>

This script iterated over each `*.lproj/Localizable.strings` of the ShareExtension target, and for each checks that all the keys and values present in that file were also present and with the same translation in the corresponding `.strings` file from the app bundle.

```ruby
#!/usr/bin/env ruby

require 'json'

### Checks that any key that was in the `*.lproj/Localizable.strings` files of `targets`
### are also present in the main app's `Resources/*.lproj/Localizable.strings`
### and with the same translation.

def dict(plist_path)
  JSON.parse(`plutil -convert json -o - #{plist_path}`)
end

targets = ['WordPressShareExtension', 'WordPressTodayWidget', 'WordPressThisWeekWidget', 'WordPressAllTimeWidget']

targets.each do |target|
  puts "== #{target} =="
  Dir["WordPress/#{target}/*.lproj/Localizable.strings"].each do |tlf|
    lproj = File.basename(File.dirname(tlf))
    next if lproj == 'Base.lproj'
    widget_strings = dict(tlf)
    app_strings = dict("WordPress/Resources/#{lproj}/Localizable.strings")
    puts " - Checking #{lproj}..."
    ok_count = 0
    widget_strings.each do |wk,wv|
      av = app_strings[wk]
      if av != wv
        puts "    ! Key #{wk} does not have the same #{lproj} translation (#{av} vs #{wv})"
      else
        ok_count += 1
      end
    end
    puts "   > #{ok_count}/#{widget_strings.count} keys matched."
  end
end
```

</details>

<details><summary>(‡) Script to find all occurrences of <tt>NSLocalizedString</tt> used in a target</summary>

This is a script I've also already used in my previous PRs, and lists all the files that are part of the "Compile Sources" build phase of the given target(s), so that I could pipe it with `xargs -0 ag 'NSLocalizedString'` to find all files used by the `ShareExtension` where this method/macro was used, and replace those with `AppLocalizedString` insterad.

```ruby
#!/usr/bin/env ruby

require 'xcodeproj'

def die!(error)
  puts error
  exit 1
end

def list(array)
  array.map { |x| " - #{x}" }.join("\n")
end

die! 'Name of a target is required as first parameter.' if ARGV.count == 0
targets_to_check = ARGV

project = Xcodeproj::Project.open('WordPress/WordPress.xcodeproj')
targets = targets_to_check.map do |name|
  project.targets.find { |t| t.name == name } \
    or die! "Can't find target #{name}. Available targets are:\n#{list(project.targets.map(&:name))}."
end

phases = targets.map do |t|
  t.build_phases.find { |p| p.is_a?(Xcodeproj::Project::Object::PBXSourcesBuildPhase) } \
    or die! "The target #{target} does not seem to have a Sources Build Phase."
end

files = phases.flat_map { |p| p.files_references.map(&:real_path) }.sort.uniq

# puts "Files compiled as part of targets #{targets_to_check.inspect}:"
# puts list(files.map(&:to_s))
print files.join("\0")
```

</details>

## To Test

 - Select the `WordPressShareExtension` scheme in Xcode
 - Clean the target, to ensure there's no leftover of `.strings` files in the target's build cache
 - Check that your simulator is in a Locale other than English. Also Edit the scheme, and under Run > Options tags, select the same locale and region for which you want to test.
 - Run the Scheme. At that point Xcode should show a popup asking you which host app you want to run (see below) to test that share extension. Personally I just chose Safari
 - Once Xcode launched Safari in your Simulator, go to a site of your choice, then hit the ⏍ share button in Safari's toolbar, and choose the WordPress action
 - In the modal that shows, check that the buttons at the top (Cancel and Next) are translated (they are barely visible at least on a phone in Light mode 😞  – but eh, 🤷 that's not that PR's problem…).
 - Then tap the Next button. The next page should show a loading screen whose copy ("Fetching sites..." in English) should also be translated in the current locale (and not appear in English)
 - Once the loading screen has finished fetching the sites and shows the next screen, check that all the copies are translated as well. Note that the list of categories of the post (on the last screen) are not translated (but I think that's expected since that's likely coming from the backend… and specific to each site?)
 - Optional: continue playing with the share extension further, maybe even actually publishing the post on one of your test site, and/or turning off your network and try to publish to trigger error messages… and check that all the flow is still translated in the locale you're testing into.

## Demo

Xcode asking you which host app to launch to test the Share Extension with:

![Xcode asking you which host app to launch to test the Share Extension with](https://user-images.githubusercontent.com/216089/145105070-fb86d22d-4ce7-4632-923f-ab4e534234ae.png)

Demo of testing the extension in French via Xcode.

https://user-images.githubusercontent.com/216089/145105594-4a407718-4de7-4989-9dfc-c9d37fe3d299.mp4


## Regression Notes

1. Potential unintended areas of impact

Translation of Share Extension in various locale.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Tested various flows of the Share Extension in simulator, with 2 locales

3. What automated tests I added (or what prevented me from doing so)

None

PR submission checklist:
